packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -224,6 +230,25 @@ export class EpochTracker implements IPersistedFileCache {
             }
             await this.checkForEpochError(error, epochFromResponse, fetchType);
             throw error;
+        }).catch((error) => {
+            // If the error is about location redirection, then we need to generate new resolved url with correct
+            // location info.
+            if (isFluidError(error) && error.errorType === DriverErrorType.locationRedirection) {
```

Comment:
If I got it right, same DriverErrorType.locationRedirection error type is used by createOdspNetworkError and by LocationRedirectionError. Callers (various entities) will assume ILocationRedirectionError shape for any error that has this errorType.
We should use different error types.

Another alternative - use DriverErrorType.fileNotFoundOrAccessDeniedError (same as for other 404 cases) but have extra mark up on error object to make it unique. That way, clients do not need to know about extra error type, and if we screw up somewhere (or if different code path propagated this error without patching), end-to-end flow would work reasonably Ok.
I like that less, as we would not see it in telemetry easily. I really want to see cases where this type of error is not handled (does not go through this path) - this will give us more data points on what to do next.

Reply:
One more consideration (which we may decide not to deal with): there are two cases:
1. We get this error on boot
2. Domain move happens halfway through existing session.

I'd assume that # 2 should be mostly solved by all SPO APis working only with driveId & itemID (i.e. not require siteUri) and this this is non-event for client. If it's an event, we need to learn about it (through telemetry). But we also know that joinSession needs to be changed to adhere to this requirement (I believe) - can you please double check and ping Gary on where this work is? I remember Robert (Gary's boss) mentioned that they were aware and needed to do this work.

Reply:
I like the `fileNotFoundOrAccessDeniedError ` ++ idea.  See [my comment below](https://github.com/microsoft/FluidFramework/pull/10581/files#r891799197) for another 2 alternatives.  I agree with the concern about using the same `errorType` for two different error shapes (classes)

Reply:
Ok I think we can do a similar thing as we already do for plumbing `serverEpoch` up:
* Update `IOdspError` to include optional `redirectLocation` property, like `serverEpoch`
  * Ok but in retrospect (I added `serverEpoch` last year) it would be better to pull them both into a separate type from `IOdspError`, call it something like `IOdspErrorAugmentations` - it would be full of all the optional bits of data we ever pull off an error object and need to refer to higher up the stack.  To use it you cast an error to it and set the property you need.
* Use a `fileNotFoundOrAccessDeniedError` and add the `redirectLocation` property 
* In `epochTracker` cast again to `IOdspErrorAugmentations` and check for the property, just like is already done for `serverEpoch`.
* Then create and throw the `LocationRedirectionError`